torch/cuda/streams.py

```diff
@@ -119,90 +102,75 @@ def __repr__(self):
             (default: ``False``)
-        blocking (bool): if ``True``, :meth:`wait` will be blocking (default: ``False``)
+        blocking (bool, optional): if ``True``, :meth:`wait` will be blocking (default: ``False``)
         interprocess (bool): if ``True``, the event can be shared between processes
             (default: ``False``)
+        _handle (bytes-like object, optional): acquired by calling Event.ipc_handle()
+            from another Event object. If given and not ``None``, all flags will be
+            ignored, and a new event will be reconstructed from the given handle.
+            Both the original and the reconstructed events must not be deconstructed
+            when calling methods (e.g., query(), synchronize()) on either event.
     """
 
-    DEFAULT = 0x0
-    BLOCKING_SYNC = 0x1
-    DISABLE_TIMING = 0x2
-    INTERPROCESS = 0x4
-
-    def __init__(self, enable_timing=False, blocking=False, interprocess=False,
-                 _handle=None):
-        flags = Event.DEFAULT
-        if not enable_timing:
-            flags |= Event.DISABLE_TIMING
-        if blocking:
-            flags |= Event.BLOCKING_SYNC
-        if interprocess:
-            flags |= Event.INTERPROCESS
-
-        ptr = ctypes.c_void_p()
-        self._cudart = cudart()
-        if _handle:
-            check_error(self._cudart.cudaIpcOpenEventHandle(ctypes.byref(ptr), _handle))
-        else:
-            check_error(self._cudart.cudaEventCreateWithFlags(ctypes.byref(ptr), ctypes.c_uint(flags)))
-        self._as_parameter_ = ptr
-
-    def __del__(self):
-        if hasattr(self, '_as_parameter_'):
-            check_error(self._cudart.cudaEventDestroy(self._as_parameter_))
-            del self._as_parameter_
+    def __new__(cls, enable_timing=False, blocking=False, interprocess=False,
+                **kwargs):
+        return super(Event, cls).__new__(
+            cls, enable_timing=enable_timing, blocking=blocking,
+            interprocess=interprocess, **kwargs)
 
     def record(self, stream=None):
-        r"""Records the event in a given stream."""
+        r"""Records the event in a given stream. Use
+        ``torch.cuda.current_stream()`` if not given. The stream's device must agree
+        with event's device."""
         if stream is None:
             stream = torch.cuda.current_stream()
-        stream.record_event(self)
+        super(Event, self).record(stream)
 
     def wait(self, stream=None):
-        r"""Makes a given stream wait for the event."""
+        r"""Makes a given stream wait for the event. Use
+        ``torch.cuda.current_stream()`` is not given."""
         if stream is None:
             stream = torch.cuda.current_stream()
-        stream.wait_event(self)
+        super(Event, self).wait(stream)
 
     def query(self):
-        r"""Checks if the event has been recorded.
+        r"""Checks if all work currently captured by event has completed. This
```

Comment:
Remove "This can be called on any device.". That's notable for CUDA's cudaEventQuery(), but we make sure that Python methods can be called from any device by setting the current device if necessary.